### PR TITLE
Issue #4265

### LINKED ISSUE #4265: [SR] Enforce some basic password constraints: no username, no email
**Description of the need**
Currently Backdrop (and D7) allow anything to be entered in the password. I think that's okay for the most part (contrib modules like Password Policy (d7-only?) can provide more constraints) but I do feel that there are a couple obvious constraints that should be enforced: **username** and **email address**.

**Proposed solution**
Don't allow users to enter username or email in their password.

---
~PR by @indigoxela: https://github.com/backdrop/backdrop/pull/3053~
~PR by @jenlamptonhttps://github.com/backdrop/backdrop/pull/3281~ (separated UI for pw length)
~PR by @herbdool https://github.com/backdrop/backdrop/pull/3215~
PR by @jenlampton https://github.com/backdrop/backdrop/pull/3282 (removed UI for pw length)

### PR DESCRIPTION
**Pull request by herbdool**
Fixes https://github.com/backdrop/backdrop-issues/issues/4265

Another approach https://github.com/backdrop/backdrop-issues/issues/4265

### CODE REVIEW

**Comment by docwilmot** [comment]
I like this approach best. But one nit: we say reject "weak" passwords, but the strength indicator never mentions weak: it says "poor", "fair", "good" and "excellent". Which of those qualifies as not weak? And we should change the wording to match that.

**Comment by herbdool** [comment]
@docwilmot good point. I was going by the user.js code which calls it "weak" but if the UI calls it "poor" then we can use that. Or we can change poor everywhere to weak.

File: core/modules/system/config/system.core.json [hunk 1]
```diff
@@ -80,6 +80,9 @@
     "user_mail_status_canceled_notify": 0,
     "user_register": "admin_only",
     "user_signatures": 0,
+    "user_password_reject_weak": 0,
+    "user_password_min_length_constraint": 0,
```
Comment (klonos) [inline core/modules/system/config/system.core.json]: I still feel that this has the risk of being confused with the min length number:

```suggestion
    "user_password_min_length_constraint": 0,
```

I preferred it with the word `_enabled`, which clearly implies that values `1`/`0` for on/off are to be expected, and these are not string length.

File: core/modules/user/user.module [hunk 2]
```diff
@@ -1044,6 +1060,117 @@ function user_account_form_validate($form, &$form_state) {
...
+    $strength = 5;
+    $messages[] = t('The password cannot be the same as the username');
+  }
+  if (backdrop_strtolower($password) == backdrop_strtolower($email)) {
+    $strength = 5;
+    $messages[] = t('The password cannot be the same as the email');
+  }
+
+  if ($strength <= 50) {
+    $form_error = implode('<br>', $messages);
```
Comment (klonos) [inline core/modules/user/user.module]: Should we use `theme('item_list', ...)` here instead? ...if `count($messages) > 1`.

File: core/modules/user/user.module [hunk 3]
```diff
@@ -1044,6 +1060,117 @@ function user_account_form_validate($form, &$form_state) {
...
+ * Adapted from https://github.com/dropbox/zxcvbn.
+ *
+ * @see Backdrop.evaluatePasswordStrength
+ *
+ * @param string $password
+ *
+ * @return int
+ *   An integer representing password strength.
+ */
+function _user_estimate_password_entropy($password) {
```
Comment (klonos) [inline core/modules/user/user.module]: Really nice 👍

File: core/modules/user/user.module [hunk 4]
```diff
@@ -2771,19 +2898,44 @@ function _user_mail_notify($op, $account, $language = NULL) {
...
+    }
     $strength_settings = array(
-      'strengthTitle' => t('Password strength: '),
-      'weak' => t('poor'),
-      'fair' => t('fair'),
-      'good' => t('good'),
-      'strong' => t('excellent'),
-      'username' => (isset($user->name) ? $user->name : ''),
+      'labels' => array(
+        'strengthTitle' => t('Password strength: '),
```
Comment (klonos) [inline core/modules/user/user.module:2872]: Symbols that are universal and especially trailing spaces should be avoided in translatable strings. This can instead be:

```suggestion
        'strengthTitle' => t('Password strength') . ': ',
```

or:

```suggestion
        'strengthTitle' => t('Password strength:') . ' ',
```

or ideally just:

```suggestion
        'strengthTitle' => t('Password strength'),
```

...and the `: ` bit can be appended to the `passwordStrengthSettings.labels.strengthTitle` in the JS.

File: core/modules/system/config/system.core.json (hunk 1 above)
Comment (herbdool) [reply core/modules/system/config/system.core.json]: Cool

File: core/modules/user/user.module (hunk 4 above)
Comment (herbdool) [reply core/modules/user/user.module:2872]: gotcha

File: core/modules/user/user.module (hunk 3 above)
Comment (herbdool) [reply core/modules/user/user.module]: Thank you. It's functionally the same as the code in user.js and here so we can use it when validating. I'm not sure if this is the best approach - eventually we might want to have the two combined in PHP and have the indicator check the strength via ajax. But for now this is probably fine.

File: core/modules/user/user.module (hunk 2 above)
Comment (herbdool) [reply core/modules/user/user.module]: I took a different approach of having each line set an error. The problem with using theme_item_list here is that it ends up with two bullets in a row.

File: core/modules/user/user.admin.inc [hunk 5]
```diff
@@ -168,6 +168,45 @@ function user_admin_settings($form, &$form_state) {
...
+    '#title' => t('Password policy'),
+    '#weight' => -1,
+  );
+  $output = '';
+  $output .= '<p>' . t('The default criteria for a strong password include checking that it is:') . '</p>';
+  $output .= '<ul><li>' . t('not the same as the username') . '</li>';
+  $output .= '<li>' . t('not the same as the e-mail address') . '</li>';
+  $output .= '<li>' . t('sufficiently complex: containing a number of lower- and uppercase letters, digits and symbols') . '</li>';
+  $output .= '<li>' . t('at least as many characters as the minimum number set below') . '</li></ul>';
+  $form['password_policy']['user_password_criteria'] = array(
```
Comment (klonos) [inline core/modules/user/user.admin.inc]: I have been advised that it is best to output things that are to be displayed to users through `theme()` functions where possible, instead of "hardcoding" html tags etc. Also, we should be rendering that as a help text, so how about this?:

```suggestion
  $output = '';
  $criteria = array();
  $criteria[] = t('not the same as the username');
  $criteria[] = t('not the same as the e-mail address');
  $criteria[] = t('sufficiently complex: containing a number of lower- and uppercase letters, digits and symbols');
  $criteria[] = t('at least as many characters as the minimum number set below');

  $output .= t('The default criteria for a strong password include checking that it is:');
  $output .= theme('item_list', array('items' => $criteria));
  $form['password_policy']['user_password_criteria'] = array(
    '#type' => 'help',
```

While at it, and since we are checking case-insensitive by default, I suggest that we change `the same as` to `similar to` instead.

File: core/modules/user/user.admin.inc [hunk 6]
```diff
@@ -168,6 +168,57 @@ function user_admin_settings($form, &$form_state) {
...
+    '#title' => t('Password policy'),
+    '#weight' => -1,
+  );
+  $output = '';
+  $criteria = array();
+  $criteria[] = t('not the same as the username');
+  $criteria[] = t('not the same as the e-mail address');
+  $criteria[] = t('sufficiently complex: containing a number of lower- and uppercase letters, digits and symbols');
+  $criteria[] = t('at least as many characters as the minimum number set below');
+
```
Comment (klonos) [inline core/modules/user/user.admin.inc]: We should remove this now. We have the "inline combo" with the number field visible all the time, so I think that this is superfluous here.

File: core/modules/user/user.admin.inc [hunk 7]
```diff
@@ -168,6 +168,57 @@ function user_admin_settings($form, &$form_state) {
...
+    '#title_display' => 'invisible',
+    '#default_value' => $config->get('user_password_min_length_enabled'),
+  );
+  $form['password_policy']['min_length']['user_password_min_length'] = array(
+    '#type' => 'number',
+    '#title' => t('Minimum password length'),
+    '#title_display' => 'invisible',
+    '#min' => 1,
+    '#max' => 64,
+    '#prefix' => t('Password must be at least'),
```
Comment (klonos) [inline core/modules/user/user.admin.inc]: Unfortunately, this is not accessible this way ^^, as the prefix of the number field does not tick/untick the checkbox. So let's change that to this then, which works:

```suggestion
    '#title' => t('Password must be at least'),
    '#default_value' => $config->get('user_password_min_length_enabled'),
  );
  $form['password_policy']['min_length']['user_password_min_length'] = array(
    '#type' => 'number',
    '#title' => t('Minimum password length'),
    '#title_display' => 'invisible',
    '#min' => 1,
    '#max' => 64,
```

File: core/modules/user/user.module [hunk 8]
```diff
@@ -1044,6 +1060,112 @@ function user_account_form_validate($form, &$form_state) {
...
+  if ($strength <= 50) {
+    form_set_error('entropy_pass', t('The password is too weak. Please consider making your password more complex.'));
+  }
+
+  $min_length_constraint = $config->get('user_password_min_length_enabled');
+  $min_length = $config->get('user_password_min_length');
+  // We need to match the JS string length method which uses UTF-16.
+  if ($min_length_constraint && (backdrop_strlen($password) < $min_length)) {
+    $strength = 5;
+    form_set_error('min_length_pass', t('The password is too short. At least @min_length_length characters required.', array('@min_length' => $min_length)));
```
Comment (klonos) [inline core/modules/user/user.module]: An underscore creeped in there 😅 :

```suggestion
    form_set_error('min_length_pass', t('The password is too short. At least @min_length length characters required.', array('@min_length' => $min_length)));
```

**Comment by quicksketch** [comment]
Closing/reopening for new sandbox.

**Comment by quicksketch** [comment]
This was ultimately superseded by https://github.com/backdrop/backdrop/pull/3282.
